linera-rpc/src/grpc/client.rs

```diff
@@ -316,6 +316,34 @@ impl ValidatorNode for GrpcClient {
         Ok(client_delegate!(self, download_certificate, hash)?.try_into()?)
     }
 
+    #[instrument(target = "grpc_client", skip_all, err, fields(address = self.address))]
+    async fn download_certificates(
+        &self,
+        hashes: Vec<CryptoHash>,
+    ) -> Result<Vec<Certificate>, NodeError> {
+        let mut missing_hashes = hashes;
+        let mut certs_collected = Vec::with_capacity(missing_hashes.len());
+        loop {
+            // Macro doesn't compile if we pass `missing_hashes.clone()` directly to `client_delegate!`.
+            let missing = missing_hashes.clone();
+            let mut received: Vec<Certificate> =
+                client_delegate!(self, download_certificates, missing)?.try_into()?;
+
+            // In the case of the server not returning any certificates, we break the loop.
+            if received.is_empty() {
+                break;
+            }
+
+            // Remove ones we've just received.
+            let received_hashes: BTreeSet<CryptoHash> =
+                received.iter().map(|cert| cert.hash()).collect();
+            missing_hashes.retain(|hash| !received_hashes.contains(hash));
```

Comment:
Alternatively, we could expect an honest validator to return the certificates in the same order as requested.

Reply:
Done in https://github.com/linera-io/linera-protocol/pull/2644/commits/b8a8dcc8c51f3d70392922de9e42a94a7e2dc49a